# **Categorizing Customer df_reviews**

## **Import Packages and Data**

In [26]:
# Import packages
import numpy as np
import pandas as pd

In [27]:
# Import data
df_reviews = pd.read_csv('/Users/kellyshreeve/Desktop/Data-Sets/imdb_reviews.tsv', 
                      sep = '\t', parse_dates=True)

In [28]:
# Data info
print(df_reviews.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 47331 entries, 0 to 47330
Data columns (total 17 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   tconst           47331 non-null  object 
 1   title_type       47331 non-null  object 
 2   primary_title    47331 non-null  object 
 3   original_title   47331 non-null  object 
 4   start_year       47331 non-null  int64  
 5   end_year         47331 non-null  object 
 6   runtime_minutes  47331 non-null  object 
 7   is_adult         47331 non-null  int64  
 8   genres           47331 non-null  object 
 9   average_rating   47329 non-null  float64
 10  votes            47329 non-null  float64
 11  review           47331 non-null  object 
 12  rating           47331 non-null  int64  
 13  sp               47331 non-null  object 
 14  pos              47331 non-null  int64  
 15  ds_part          47331 non-null  object 
 16  idx              47331 non-null  int64  
dtypes: float64(2

In [29]:
# Data head
display(df_reviews.sample(15))

tconst    title_type                      primary_title  \
39080  tt0098849  tvMiniSeries  The Life and Loves of a She-Devil   
36056  tt0228183         movie                     The Dark Power   
38556  tt0115669         movie                  The King of Masks   
14476  tt0109900         movie                           Giorgino   
8026   tt0074515         movie                      Coup de Grâce   
10419  tt0275309         movie                 Dogtown and Z-Boys   
11812  tt0082332         movie                    Enter the Ninja   
11217  tt0123816      tvSeries              Earth: Final Conflict   
39273  tt0025420         movie                    The Lost Jungle   
43258  tt0063683         movie               Therese and Isabelle   
32103  tt0350146         video                         SpiderBabe   
21544  tt0476614  tvMiniSeries       Man to Man with Dean Learner   
24848  tt0273907         video           Octopus 2: River of Fear   
41214  tt0093894         movie                    The Running Man   
28234  tt0787505         video                          Rest Stop   

                          original_title  start_year end_year runtime_minutes  \
39080  The Life and Loves of a She-Devil        1986     1986             229   
36056                     The Dark Power        1985       \N              87   
38556                          Bian Lian        1996       \N              91   
14476                           Giorgino        1994       \N             177   
8026                       Der Fangschuß        1976       \N              97   
10419                 Dogtown and Z-Boys        2001       \N              91   
11812                    Enter the Ninja        1981       \N              99   
11217              Earth: Final Conflict        1997     2002              60   
39273                    The Lost Jungle        1934       \N             243   
43258               Therese and Isabelle        1968       \N             118   
32103                         SpiderBabe        2003       \N              89   
21544       Man to Man with Dean Learner        2006       \N              \N   
24848           Octopus 2: River of Fear        2001       \N              91   
41214                    The Running Man        1987       \N             101   
28234                          Rest Stop        2006       \N              80   

       is_adult                   genres  average_rating     votes  \
39080         0                    Drama             8.0     540.0   
36056         0                   Horror             3.4     531.0   
38556         0                    Drama             7.7    3056.0   
14476         0   Adventure,Drama,Horror             7.4     879.0   
8026          0                Drama,War             7.0    1125.0   
10419         0        Documentary,Sport             7.7   11993.0   
11812         0       Action,Crime,Drama             5.3    4020.0   
11217         0     Action,Drama,Mystery             6.2    3759.0   
39273         0   Action,Adventure,Drama             4.8     193.0   
43258         0            Drama,Romance             6.2     545.0   
32103         0  Adventure,Comedy,Sci-Fi             4.3     806.0   
21544         0         Comedy,Talk-Show             7.7    1389.0   
24848         0  Action,Adventure,Horror             2.9     905.0   
41214         0   Action,Sci-Fi,Thriller             6.7  137663.0   
28234         0  Horror,Romance,Thriller             4.6   10678.0   

                                                  review  rating   sp  pos  \
39080  I'm a big fan of Patricia Hodge and Mariam Mar...       1  neg    0   
36056  A total and utter travesty of a movie.'Dark po...       1  neg    0   
38556  The movie is truly poignant, unique and uplift...      10  pos    1   
14476   Giorgino is a strange, dark, obsessive object...      10  pos    1   
8026   The film begins in Latvia just after WWI. Bein...       3  neg    0   
10419  I remember w

### Initial data observations

The dataset contains 47331 observations and 16 columns representing movie df_df_df_reviews and their classification, positive or negative. Runtime_minutes needs to be converted to int64. There are no missing values.

## **Prepare Data**

### Fix data types

In [30]:
# Check values of runtime
display(sorted(df_reviews['runtime_minutes'].unique()))

['1',
 '10',
 '100',
 '101',
 '102',
 '103',
 '104',
 '105',
 '106',
 '107',
 '108',
 '109',
 '11',
 '110',
 '111',
 '112',
 '113',
 '114',
 '1140',
 '115',
 '116',
 '117',
 '118',
 '119',
 '12',
 '120',
 '121',
 '122',
 '123',
 '124',
 '125',
 '126',
 '127',
 '128',
 '129',
 '13',
 '130',
 '131',
 '132',
 '133',
 '134',
 '135',
 '136',
 '137',
 '138',
 '139',
 '14',
 '140',
 '141',
 '142',
 '143',
 '144',
 '145',
 '146',
 '147',
 '148',
 '149',
 '15',
 '150',
 '151',
 '152',
 '153',
 '154',
 '155',
 '156',
 '157',
 '159',
 '16',
 '160',
 '161',
 '162',
 '163',
 '164',
 '165',
 '166',
 '167',
 '168',
 '169',
 '17',
 '170',
 '171',
 '172',
 '173',
 '174',
 '175',
 '176',
 '177',
 '178',
 '179',
 '18',
 '180',
 '181',
 '183',
 '185',
 '186',
 '188',
 '189',
 '19',
 '190',
 '192',
 '194',
 '197',
 '199',
 '2',
 '20',
 '200',
 '201',
 '202',
 '204',
 '208',
 '21',
 '210',
 '211',
 '212',
 '215',
 '216',
 '217',
 '219',
 '22',
 '222',
 '225',
 '229',
 '23',
 '230',
 '232',
 '233',
 '239',
 

Missing values for runtime were entered as '\\N', which is causing the data to be mis-typed as object. Changing '\\N' to np.nan will allow the variable to be converted to to float64.

In [31]:
# Replace '\\N' with np.nan
df_reviews['runtime_minutes'] = df_reviews['runtime_minutes'].replace('\\N', np.nan)

# Convert runtime to int64
df_reviews['runtime_minutes'] = df_reviews['runtime_minutes'].astype('float')

# Display info
print(df_reviews.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 47331 entries, 0 to 47330
Data columns (total 17 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   tconst           47331 non-null  object 
 1   title_type       47331 non-null  object 
 2   primary_title    47331 non-null  object 
 3   original_title   47331 non-null  object 
 4   start_year       47331 non-null  int64  
 5   end_year         47331 non-null  object 
 6   runtime_minutes  46843 non-null  float64
 7   is_adult         47331 non-null  int64  
 8   genres           47331 non-null  object 
 9   average_rating   47329 non-null  float64
 10  votes            47329 non-null  float64
 11  review           47331 non-null  object 
 12  rating           47331 non-null  int64  
 13  sp               47331 non-null  object 
 14  pos              47331 non-null  int64  
 15  ds_part          47331 non-null  object 
 16  idx              47331 non-null  int64  
dtypes: float64(3

Runtime_minutes is converted to float.

### Check for duplicates

In [32]:
# Check for full duplicates
duplicates_count = df_reviews.duplicated().sum()

print(f'Number of full duplicates: {duplicates_count}')

Number of full duplicates: 0


There are no fully duplicate rows.

In [33]:
# Check for review duplicates
review_duplicates = df_reviews['review'].duplicated().sum()

print(f'Number of duplicated review texts: {review_duplicates}')

Number of duplicated review texts: 91


There are 91 duplicated df_reviews. Print rows to further inspect the duplication.

In [34]:
# Print duplicated rows
duplicate_review_text = df_reviews['review'].duplicated(keep=False)

display(df_reviews[duplicate_review_text])

tconst    title_type         primary_title        original_title  \
1368   tt0081949  tvMiniSeries     A Town Like Alice     A Town Like Alice   
1370   tt0081949  tvMiniSeries     A Town Like Alice     A Town Like Alice   
1629   tt0098994         movie  After Dark, My Sweet  After Dark, My Sweet   
1630   tt0098994         movie  After Dark, My Sweet  After Dark, My Sweet   
1753   tt0086662      tvSeries               Airwolf               Airwolf   
...          ...           ...                   ...                   ...   
47175  tt0463392         movie         Zombie Nation         Zombie Nation   
47194  tt0463392         movie         Zombie Nation         Zombie Nation   
47288  tt0972534      tvSeries                iCarly                iCarly   
47290  tt0972534      tvSeries                iCarly                iCarly   
47292  tt0972534      tvSeries                iCarly                iCarly   

       start_year end_year  runtime_minutes  is_adult  \
1368         1981     1981            301.0         0   
1370         1981     1981            301.0         0   
1629         1990       \N            114.0         0   
1630         1990       \N            114.0         0   
1753         1984     1986             60.0         0   
...           ...      ...              ...       ...   
47175        2004       \N             81.0         0   
47194        2004       \N             81.0         0   
47288        2007     2012             30.0         0   
47290        2007     2012             30.0         0   
47292        2007     2012             30.0         0   

                        genres  average_rating    votes  \
1368         Drama,Romance,War             8.3    849.0   
1370         Drama,Romance,War             8.3    849.0   
1629       Crime,Drama,Mystery             6.6   3098.0   
1630       Crime,Drama,Mystery             6.6   3098.0   
1753   Action,Adventure,Sci-Fi             6.7   9592.0   
...                        ...             ...      ...   
47175                   Horror             2.0   7850.0   
47194                   Horror             2.0   7850.0   
47288    Comedy,Family,Romance             6.6  31143.0   
47290    Comedy,Family,Romance             6.6  31143.0   
47292    Comedy,Family,Romance             6.6  31143.0   

                                                  review  rating   sp  pos  \
1368   I have seen this film at least 100 times and I...      10  pos    1   
1370   I have seen this film at least 100 times and I...      10  pos    1   
1629   THE JIST: See something else. This film was hi...       7  pos    1   
1630   THE JIST: See something else. This film was hi...       7  pos    1   
1753   I used to watch this show when I was a little ...       9  pos    1   
...                                                  ...     ...  ...  ...   
47175  German filmmaker Ulli Lommel has managed a tas...       1  neg    0   
47194  German filmmaker Ulli Lommel has managed a tas...       1  neg    0   
47288  Nickelodeon has gone down the toilet. They hav...       1  neg    0   
47290  Nickelodeon has gone down the toilet. They hav...       1  neg    0   
47292  Nickelodeon has gone down the toilet. They hav...       1  neg    0   

      ds_part    idx  
1368     test   9765  
1370     test   9768  
1629    train   7728  
1630    train   7729  
1753    train   6769  
...       ...    ...  
47175   train  10089  
47194   train  10108  
47288    test   5652  
47290    test   5655  
47292    test   5658  

[173 rows x 17 columns]

The rows are fully duplicated other than the idx index variable. Duplicates will be dropped from the dataset.

In [35]:
# Drop duplicates
reivews = df_reviews.drop_duplicates(subset=['review'], inplace=True)

duplicates_new = df_reviews['review'].duplicated().sum()

print(f'Updated number of duplicates: {duplicates_new}')

Updated number of duplicates: 0


Duplicates have been removed from the dataset.

## Missing values

In [42]:
# Replace all '\N' with np.nan
df_reviews = df_reviews.replace('\\N', np.nan)

In [48]:
# Calculate number and percent of missing values by columns
missing = df_reviews.isna().sum().reset_index().rename(columns=
                                                       {'index':'column',
                                                        0:'count'})

missing['percent'] = ((missing['count'] / len(df_reviews)) * 100).round(2)

print('Count and Percent Missing Values')
display(missing)

Count and Percent Missing Values


column  count  percent
0            tconst      0      0.0
1        title_type      0      0.0
2     primary_title      0      0.0
3    original_title      0      0.0
4        start_year      0      0.0
5          end_year  44586     95.4
6   runtime_minutes      0      0.0
7          is_adult      0      0.0
8            genres      0      0.0
9    average_rating      0      0.0
10            votes      0      0.0
11           review      0      0.0
12           rating      0      0.0
13               sp      0      0.0
14              pos      0      0.0
15          ds_part      0      0.0
16              idx      0      0.0

End_year is missing over 95% of values, runtime_minutes is missing 1% of valeus, and genres, average_rating, and votes are all missing less than 1% of values. Because of the high number missing in end_year, this variable will be dropped from the analysis. The number missing in other variables is so low that imputation isn't necessary. The rows missing information will be dropped from the dataset.

In [49]:
# Drop rows missing values in runtime, genre, rating, or votes
df_reviews = df_reviews.dropna(subset=['runtime_minutes', 'genres', 'average_rating', 'votes'])

missing_dropped = df_reviews.isna().sum().reset_index().rename(columns=
                                                       {'index':'column',
                                                        0:'count'})

missing_dropped['percent'] = ((missing_dropped['count'] / len(df_reviews)) * 100).round(2)

print('Count and Percent Missing after dropping missing rows')
display(missing_dropped)

Count and Percent Missing after dropping missing rows


column  count  percent
0            tconst      0      0.0
1        title_type      0      0.0
2     primary_title      0      0.0
3    original_title      0      0.0
4        start_year      0      0.0
5          end_year  44586     95.4
6   runtime_minutes      0      0.0
7          is_adult      0      0.0
8            genres      0      0.0
9    average_rating      0      0.0
10            votes      0      0.0
11           review      0      0.0
12           rating      0      0.0
13               sp      0      0.0
14              pos      0      0.0
15          ds_part      0      0.0
16              idx      0      0.0

There are no more missing values, after dropping rows missing in the runtime, genre, rating, and votes subest. End_year will be left out of analyses.

### Prepare data conclusion

Duplicates and missing values have been dropped from the dataset. The data is ready for analysis.

## **Exploratory data analysis**

In [56]:
movie_count = df_reviews['primary_title'].unique().count()

print(movie_count)

AttributeError: 'numpy.ndarray' object has no attribute 'count'